### PR TITLE
[Lesson 16] Add waiters and implement new test:

### PR DESCRIPTION
**Pull request by TatyanaDovzhich**
- Add implicitly wait on driver level
- Add explicitly waits on base page level
- Add decorator to handle "magic-strange-weird" Sign Up button
- Added test for post creation with related constants and pages

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -17,4 +20,29 @@ def driver():
 def start_page(driver):
     """Creates start page object"""
     driver.get(BaseConstants.URL)
+    driver.implicity_wait(1)
     return StartPage(driver)
+
+
+@pytest.fixture()
+def hello_page(driver):
```
Comment (QALightLecorPython) [inline conftest.py]: It is not working like this. You cannot just create any page. Only start page can be created from driver by opening it, all other pages are have to be crated by navigating to them under the page objects

**Review by QALightLecorPython — CHANGES_REQUESTED** [review]

File: conftest.py (hunk 1 above)
Comment (TatyanaDovzhich) [reply conftest.py]: Fixings are:
-  delete unnecessary wrong drivers at conftest.py;
- create waiter for Sign In button at the Start Page;
- create waiters for Create Post and Sign Out button at the header;
- create waiter for Save New Post button at the Create Post page.

File: pages/create_post_page.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
-import logging
+from enum import verify
```
Comment (QALightLecorPython) [inline pages/create_post_page.py:1]: what is this ?

File: pages/create_post_page.py [hunk 3]
```diff
@@ -10,12 +12,26 @@ class CreatePostPage(BasePage):
...
+        """Click on Save Post button until it disappear"""
+        self.click(self.const.SAVE_POST_BUTTON_XPATH)
+        assert not self.is_element_exists(self.const.SAVE_POST_BUTTON_XPATH), "Save Post button didn't disappear"
 
     def create_post(self, title, body):
         """Create post using provided values"""
         self.fill_field(xpath=self.const.TITLE_INPUT_XPATH, value=title)
         self.fill_field(xpath=self.const.BODY_AREA_XPATH, value=body)
+        # Click on Save Post button
+        if verify:
```
Comment (QALightLecorPython) [inline pages/create_post_page.py:28]: I think you don't need this logic here. But if I wrong, you should add additional argument to the method and pass it from the test

File: pages/create_post_page.py [hunk 4]
```diff
@@ -10,12 +12,26 @@ class CreatePostPage(BasePage):
     def __init__(self, driver):
         super().__init__(driver)
         self.const = CreatePostPageConsts
-        self.log = logging.getLogger("[CreatePostPage]")
+        self.header = Header(self.driver)
+
+    @wait_until_ok(timeout=3, period=0.5)
+    def click_and_validate_save_post_button(self):
```
Comment (QALightLecorPython) [inline pages/create_post_page.py:18]: Do you have a trouble with this button or you add this `just in case` ?

File: pages/header.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    def sign_out_button(self, username, password):
+        """Click on Sign Out button"""
+        if verify:
+            self.click_and_validate_sign_out_button()
+        else:
+            self.click(self.const.SIGN_OUT_BUTTON_XPATH)
+
+    # @log_wrapper
+    def navigate_to_create_post(self):
```
Comment (QALightLecorPython) [inline pages/header.py:44]: This method should return CreatePostPage

File: pages/header.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Header(BasePage):
+
+    def __init__(self, driver):
+        super().__init__(driver)
+        self.const = HeaderConsts
+
+    @wait_until_ok(timeout=3, period=0.5)
+    def click_and_validate_create_post_button(self):
```
Comment (QALightLecorPython) [inline pages/header.py:18]: Why do you need this verification ?
Anyway this button won't disappear as header is always exists.

File: pages/header.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    def create_post_button(self, username, password):
+        """Click on Create Post button"""
+        if verify:
+            self.click_and_validate_create_post_button()
+        else:
+            self.click(self.const.CREATE_POST_BUTTON_XPATH)
+
+    @wait_until_ok(timeout=3, period=0.5)
+    def click_and_validate_sign_out_button(self):
```
Comment (QALightLecorPython) [inline pages/header.py:31]: Why do you need this ? You already have `navigate_to_start_page`

**Review by QALightLecorPython — CHANGES_REQUESTED** [review]
I don't get the reason of some changes, please fix or let me know why did you do them )

**Pull request merged**